### PR TITLE
C API: Safer function pointer casting

### PR DESCRIPTION
**Pull request by jlesquembre**
# Motivation

See https://github.com/NixOS/nix/pull/8699#discussion_r1554312181

Casting a function pointer to `void*` is undefined behavior in the C spec, since there are platforms with different sizes for these two kinds of pointers. A safe alternative might be `void (*callback)()`

# Context
Stabilize the C interface introduced in https://github.com/NixOS/nix/pull/8699 

Milestone: https://github.com/NixOS/nix/milestone/52

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: src/libexpr-c/nix_api_value.cc [hunk 1]
```diff
@@ -15,9 +15,9 @@
 #include "value/context.hh"
 
 #ifdef HAVE_BOEHMGC
-# include "gc/gc.h"
-# define GC_INCLUDE_NEW 1
-# include "gc_cpp.h"
+#include "gc/gc.h"
+#define GC_INCLUDE_NEW 1
+#include "gc_cpp.h"
```
Comment (Ericson2314) [inline src/libexpr-c/nix_api_value.cc:20]: Please don't do this, the indentation between `ifdef` and `endif` helps legibility.

**Comment by Ericson2314** [comment]
`void (const char * start, unsigned int n, void * user_data)` seems worthy of a typedef?

File: src/libexpr-c/nix_api_value.cc (hunk 1 above)
Comment (Ericson2314) [reply src/libexpr-c/nix_api_value.cc:20]: Did `clang-format` with our settings do this by any chance?

Comment (Ericson2314) [reply src/libexpr-c/nix_api_value.cc:20]: If so, https://github.com/NixOS/nix/pull/10493 will ensure it won't anymore.

Comment (jlesquembre) [reply src/libexpr-c/nix_api_value.cc:20]: yep, that was caused by `cland-format` settings (my editor auto-formats it on save)
I'm going to revert it

**Review by Ericson2314 — APPROVED** [review]

**Pull request merged**